From: Simon McVittie <smcv@debian.org>
Date: Sat, 21 Jan 2012 17:19:30 +0000
Subject: Don't build SSE-only software rasterizer on non-x86 CPUs

Compilers for non-x86 platforms don't understand -msse or -msse2, and
these files aren't usable on non-x86 anyway.

Origin: vendor, Debian
Forwarded: no
---
 dpsoftrast.h  |  4 ++++
 gl_backend.c  | 56 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 gl_rmain.c    | 19 +++++++++++++++++++
 gl_textures.c | 18 ++++++++++++++++++
 makefile      | 12 ++++++++++++
 makefile.inc  | 16 +++++++++-------
 quakedef.h    |  5 +++++
 vid_glx.c     |  8 +++++++-
 vid_sdl.c     |  6 +++++-
 vid_shared.c  |  4 +++-
 vid_wgl.c     |  6 +++++-
 11 files changed, 143 insertions(+), 11 deletions(-)

diff --git a/dpsoftrast.h b/dpsoftrast.h
index 40c48b1..d8c278d 100644
--- a/dpsoftrast.h
+++ b/dpsoftrast.h
@@ -33,6 +33,7 @@ typedef enum DPSOFTRAST_TEXTURE_FILTER_e
 }
 DPSOFTRAST_TEXTURE_FILTER;
 
+#ifdef HAVE_DPSOFTRAST
 int DPSOFTRAST_Init(int width, int height, int numthreads, int interlace, unsigned int *colorpixels, unsigned int *depthpixels);
 void DPSOFTRAST_Shutdown(void);
 void DPSOFTRAST_Flush(void);
@@ -74,6 +75,7 @@ void DPSOFTRAST_SetVertexPointer(const float *vertex3f, size_t stride);
 void DPSOFTRAST_SetColorPointer(const float *color4f, size_t stride);
 void DPSOFTRAST_SetColorPointer4ub(const unsigned char *color4ub, size_t stride);
 void DPSOFTRAST_SetTexCoordPointer(int unitnum, int numcomponents, size_t stride, const float *texcoordf);
+#endif
 
 typedef enum gl20_texunit_e
 {
@@ -322,6 +324,7 @@ typedef enum DPSOFTRAST_UNIFORM_e
 }
 DPSOFTRAST_UNIFORM;
 
+#ifdef HAVE_DPSOFTRAST
 void DPSOFTRAST_SetShader(int mode, dpuint64 permutation, int exactspecularmath);
 #define DPSOFTRAST_Uniform1f(index, v0) DPSOFTRAST_Uniform4f(index, v0, 0, 0, 0)
 #define DPSOFTRAST_Uniform2f(index, v0, v1) DPSOFTRAST_Uniform4f(index, v0, v1, 0, 0)
@@ -332,5 +335,6 @@ void DPSOFTRAST_UniformMatrix4fv(DPSOFTRAST_UNIFORM index, int arraysize, int tr
 void DPSOFTRAST_Uniform1i(DPSOFTRAST_UNIFORM index, int i0);
 
 void DPSOFTRAST_DrawTriangles(int firstvertex, int numvertices, int numtriangles, const int *element3i, const unsigned short *element3s);
+#endif
 
 #endif // DPSOFTRAST_H
diff --git a/gl_backend.c b/gl_backend.c
index fa8fc68..bbf1a3c 100644
--- a/gl_backend.c
+++ b/gl_backend.c
@@ -653,7 +653,9 @@ void GL_Finish(void)
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		DPSOFTRAST_Finish();
+#endif
 		break;
 	}
 }
@@ -1204,7 +1206,9 @@ void R_SetViewport(const r_viewport_t *v)
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		DPSOFTRAST_Viewport(v->x, v->y, v->width, v->height);
+#endif
 		break;
 	case RENDERPATH_GL20:
 	case RENDERPATH_GLES2:
@@ -1521,6 +1525,7 @@ void R_Mesh_SetRenderTargets(int fbo, rtexture_t *depthtexture, rtexture_t *colo
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		if (fbo)
 		{
 			int width, height;
@@ -1534,6 +1539,7 @@ void R_Mesh_SetRenderTargets(int fbo, rtexture_t *depthtexture, rtexture_t *colo
 		}
 		else
 			DPSOFTRAST_SetRenderTargets(vid.width, vid.height, vid.softdepthpixels, vid.softpixels, NULL, NULL, NULL);
+#endif
 		break;
 	}
 }
@@ -1689,6 +1695,7 @@ static void GL_Backend_ResetState(void)
 #endif
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		DPSOFTRAST_ColorMask(1,1,1,1);
 		DPSOFTRAST_BlendFunc(gl_state.blendfunc1, gl_state.blendfunc2);
 		DPSOFTRAST_CullFace(gl_state.cullface);
@@ -1697,6 +1704,7 @@ static void GL_Backend_ResetState(void)
 		DPSOFTRAST_PolygonOffset(gl_state.polygonoffset[0], gl_state.polygonoffset[1]);
 		DPSOFTRAST_SetRenderTargets(vid.width, vid.height, vid.softdepthpixels, vid.softpixels, NULL, NULL, NULL);
 		DPSOFTRAST_Viewport(0, 0, vid.width, vid.height);
+#endif
 		break;
 	case RENDERPATH_GL20:
 	case RENDERPATH_GLES2:
@@ -1887,7 +1895,9 @@ void GL_BlendFunc(int blendfunc1, int blendfunc2)
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_BlendFunc(gl_state.blendfunc1, gl_state.blendfunc2);
+#endif
 			break;
 		}
 	}
@@ -1920,7 +1930,9 @@ void GL_DepthMask(int state)
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_DepthMask(gl_state.depthmask);
+#endif
 			break;
 		}
 	}
@@ -1960,7 +1972,9 @@ void GL_DepthTest(int state)
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_DepthTest(gl_state.depthtest);
+#endif
 			break;
 		}
 	}
@@ -1993,7 +2007,9 @@ void GL_DepthFunc(int state)
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_DepthFunc(gl_state.depthfunc);
+#endif
 			break;
 		}
 	}
@@ -2039,7 +2055,9 @@ void GL_DepthRange(float nearfrac, float farfrac)
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_DepthRange(gl_state.depthrange[0], gl_state.depthrange[1]);
+#endif
 			break;
 		}
 	}
@@ -2198,7 +2216,9 @@ void GL_PolygonOffset(float planeoffset, float depthoffset)
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_PolygonOffset(gl_state.polygonoffset[0], gl_state.polygonoffset[1]);
+#endif
 			break;
 		}
 	}
@@ -2236,7 +2256,9 @@ void GL_SetMirrorState(qboolean state)
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_CullFace(gl_state.cullface);
+#endif
 			break;
 		}
 	}
@@ -2313,12 +2335,14 @@ void GL_CullFace(int state)
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		if (gl_state.cullface != state)
 		{
 			gl_state.cullface = state;
 			gl_state.cullfaceenable = state != GL_NONE ? true : false;
 			DPSOFTRAST_CullFace(gl_state.cullface);
 		}
+#endif
 		break;
 	}
 }
@@ -2422,7 +2446,9 @@ void GL_ColorMask(int r, int g, int b, int a)
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_ColorMask(r, g, b, a);
+#endif
 			break;
 		}
 	}
@@ -2453,7 +2479,9 @@ void GL_Color(float cr, float cg, float cb, float ca)
 			// no equivalent in D3D
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_Color4f(cr, cg, cb, ca);
+#endif
 			break;
 		case RENDERPATH_GL20:
 		case RENDERPATH_GLES2:
@@ -2495,7 +2523,9 @@ void GL_Scissor (int x, int y, int width, int height)
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		DPSOFTRAST_Scissor(x, y, width, height);
+#endif
 		break;
 	}
 }
@@ -2531,7 +2561,9 @@ void GL_ScissorTest(int state)
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_ScissorTest(gl_state.scissortest);
+#endif
 			break;
 		}
 	}
@@ -2587,10 +2619,12 @@ void GL_Clear(int mask, const float *colorvalue, float depthvalue, int stencilva
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		if (mask & GL_COLOR_BUFFER_BIT)
 			DPSOFTRAST_ClearColor(colorvalue[0], colorvalue[1], colorvalue[2], colorvalue[3]);
 		if (mask & GL_DEPTH_BUFFER_BIT)
 			DPSOFTRAST_ClearDepth(depthvalue);
+#endif
 		break;
 	}
 }
@@ -2672,7 +2706,9 @@ void GL_ReadPixelsBGRA(int x, int y, int width, int height, unsigned char *outpi
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		DPSOFTRAST_GetPixelsBGRA(x, y, width, height, outpixels);
+#endif
 		break;
 	}
 }
@@ -3264,7 +3300,9 @@ void R_Mesh_Draw(int firstvertex, int numvertices, int firsttriangle, int numtri
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_DrawTriangles(firstvertex, numvertices, numtriangles, element3i, element3s);
+#endif
 			break;
 		case RENDERPATH_GLES1:
 		case RENDERPATH_GLES2:
@@ -3841,7 +3879,9 @@ void R_Mesh_CopyToTexture(rtexture_t *tex, int tx, int ty, int sx, int sy, int w
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		DPSOFTRAST_CopyRectangleToTexture(tex->texnum, 0, tx, ty, sx, sy, width, height);
+#endif
 		break;
 	}
 }
@@ -4020,6 +4060,7 @@ void R_Mesh_TexBind(unsigned int unitnum, rtexture_t *tex)
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		if (!tex)
 		{
 			tex = r_texture_white;
@@ -4032,6 +4073,7 @@ void R_Mesh_TexBind(unsigned int unitnum, rtexture_t *tex)
 			return;
 		unit->texture = tex;
 		DPSOFTRAST_SetTexture(unitnum, texnum);
+#endif
 		break;
 	}
 }
@@ -4383,6 +4425,7 @@ void R_Mesh_PrepareVertices_Vertex3f(int numvertices, const float *vertex3f, con
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		DPSOFTRAST_SetVertexPointer(vertex3f, sizeof(float[3]));
 		DPSOFTRAST_SetColorPointer(NULL, 0);
 		DPSOFTRAST_SetTexCoordPointer(0, 2, sizeof(float[2]), NULL);
@@ -4390,6 +4433,7 @@ void R_Mesh_PrepareVertices_Vertex3f(int numvertices, const float *vertex3f, con
 		DPSOFTRAST_SetTexCoordPointer(2, 2, sizeof(float[2]), NULL);
 		DPSOFTRAST_SetTexCoordPointer(3, 2, sizeof(float[2]), NULL);
 		DPSOFTRAST_SetTexCoordPointer(4, 2, sizeof(float[2]), NULL);
+#endif
 		break;
 	}
 }
@@ -4483,6 +4527,7 @@ void R_Mesh_PrepareVertices_Generic_Arrays(int numvertices, const float *vertex3
 	case RENDERPATH_D3D11:
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		DPSOFTRAST_SetVertexPointer(vertex3f, sizeof(float[3]));
 		DPSOFTRAST_SetColorPointer(color4f, sizeof(float[4]));
 		DPSOFTRAST_SetTexCoordPointer(0, 2, sizeof(float[2]), texcoord2f);
@@ -4490,6 +4535,7 @@ void R_Mesh_PrepareVertices_Generic_Arrays(int numvertices, const float *vertex3
 		DPSOFTRAST_SetTexCoordPointer(2, 2, sizeof(float[2]), NULL);
 		DPSOFTRAST_SetTexCoordPointer(3, 2, sizeof(float[2]), NULL);
 		DPSOFTRAST_SetTexCoordPointer(4, 2, sizeof(float[2]), NULL);
+#endif
 		return;
 	}
 
@@ -4602,6 +4648,7 @@ void R_Mesh_PrepareVertices_Generic(int numvertices, const r_vertexgeneric_t *ve
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		DPSOFTRAST_SetVertexPointer(vertex->vertex3f, sizeof(*vertex));
 		DPSOFTRAST_SetColorPointer(vertex->color4f, sizeof(*vertex));
 		DPSOFTRAST_SetTexCoordPointer(0, 2, sizeof(*vertex), vertex->texcoord2f);
@@ -4609,6 +4656,7 @@ void R_Mesh_PrepareVertices_Generic(int numvertices, const r_vertexgeneric_t *ve
 		DPSOFTRAST_SetTexCoordPointer(2, 2, sizeof(*vertex), NULL);
 		DPSOFTRAST_SetTexCoordPointer(3, 2, sizeof(*vertex), NULL);
 		DPSOFTRAST_SetTexCoordPointer(4, 2, sizeof(*vertex), NULL);
+#endif
 		break;
 	}
 }
@@ -4714,6 +4762,7 @@ void R_Mesh_PrepareVertices_Mesh_Arrays(int numvertices, const float *vertex3f,
 	case RENDERPATH_D3D11:
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		DPSOFTRAST_SetVertexPointer(vertex3f, sizeof(float[3]));
 		DPSOFTRAST_SetColorPointer(color4f, sizeof(float[4]));
 		DPSOFTRAST_SetTexCoordPointer(0, 2, sizeof(float[2]), texcoordtexture2f);
@@ -4721,6 +4770,7 @@ void R_Mesh_PrepareVertices_Mesh_Arrays(int numvertices, const float *vertex3f,
 		DPSOFTRAST_SetTexCoordPointer(2, 3, sizeof(float[3]), tvector3f);
 		DPSOFTRAST_SetTexCoordPointer(3, 3, sizeof(float[3]), normal3f);
 		DPSOFTRAST_SetTexCoordPointer(4, 2, sizeof(float[2]), texcoordlightmap2f);
+#endif
 		return;
 	}
 
@@ -4844,6 +4894,7 @@ void R_Mesh_PrepareVertices_Mesh(int numvertices, const r_vertexmesh_t *vertex,
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		DPSOFTRAST_SetVertexPointer(vertex->vertex3f, sizeof(*vertex));
 		DPSOFTRAST_SetColorPointer(vertex->color4f, sizeof(*vertex));
 		DPSOFTRAST_SetTexCoordPointer(0, 2, sizeof(*vertex), vertex->texcoordtexture2f);
@@ -4851,6 +4902,7 @@ void R_Mesh_PrepareVertices_Mesh(int numvertices, const r_vertexmesh_t *vertex,
 		DPSOFTRAST_SetTexCoordPointer(2, 3, sizeof(*vertex), vertex->tvector3f);
 		DPSOFTRAST_SetTexCoordPointer(3, 3, sizeof(*vertex), vertex->normal3f);
 		DPSOFTRAST_SetTexCoordPointer(4, 2, sizeof(*vertex), vertex->texcoordlightmap2f);
+#endif
 		break;
 	}
 }
@@ -4880,7 +4932,9 @@ void GL_BlendEquationSubtract(qboolean negated)
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_BlendSubtract(true);
+#endif
 			break;
 		}
 	}
@@ -4907,7 +4961,9 @@ void GL_BlendEquationSubtract(qboolean negated)
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_BlendSubtract(false);
+#endif
 			break;
 		}
 	}
diff --git a/gl_rmain.c b/gl_rmain.c
index 36a45b0..38f2eaa 100644
--- a/gl_rmain.c
+++ b/gl_rmain.c
@@ -1909,6 +1909,7 @@ void R_SetupShader_SetPermutationHLSL(unsigned int mode, dpuint64 permutation)
 }
 #endif
 
+#ifdef HAVE_DPSOFTRAST
 static void R_SetupShader_SetPermutationSoft(unsigned int mode, dpuint64 permutation)
 {
 	DPSOFTRAST_SetShader(mode, permutation, r_shadow_glossexact.integer);
@@ -1916,6 +1917,7 @@ static void R_SetupShader_SetPermutationSoft(unsigned int mode, dpuint64 permuta
 	DPSOFTRAST_UniformMatrix4fv(DPSOFTRAST_UNIFORM_ModelViewMatrixM1, 1, false, gl_modelview16f);
 	DPSOFTRAST_Uniform1f(DPSOFTRAST_UNIFORM_ClientTime, cl.time);
 }
+#endif
 
 void R_GLSL_Restart_f(void)
 {
@@ -2085,9 +2087,11 @@ void R_SetupShader_Generic(rtexture_t *first, rtexture_t *second, int texturemod
 		R_Mesh_TexMatrix(0, NULL);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		R_SetupShader_SetPermutationSoft(SHADERMODE_GENERIC, permutation);
 		R_Mesh_TexBind(GL20TU_FIRST , first );
 		R_Mesh_TexBind(GL20TU_SECOND, second);
+#endif
 		break;
 	}
 }
@@ -2138,7 +2142,9 @@ void R_SetupShader_DepthOrShadow(qboolean notrippy, qboolean depthrgb, qboolean
 		R_Mesh_TexBind(0, 0);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		R_SetupShader_SetPermutationSoft(SHADERMODE_DEPTH_OR_SHADOW, permutation);
+#endif
 		break;
 	}
 }
@@ -2841,6 +2847,7 @@ void R_SetupShader_Surface(const float rtlightambient[3], const float rtlightdif
 	case RENDERPATH_GLES1:
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		RSurf_PrepareVerticesForBatch(BATCHNEED_ARRAY_VERTEX | BATCHNEED_ARRAY_NORMAL | BATCHNEED_ARRAY_VECTOR | (rsurface.modellightmapcolor4f ? BATCHNEED_ARRAY_VERTEXCOLOR : 0) | BATCHNEED_ARRAY_TEXCOORD | (rsurface.uselightmaptexture ? BATCHNEED_ARRAY_LIGHTMAP : 0) | BATCHNEED_ALLOWMULTIDRAW, texturenumsurfaces, texturesurfacelist);
 		R_Mesh_PrepareVertices_Mesh_Arrays(rsurface.batchnumvertices, rsurface.batchvertex3f, rsurface.batchsvector3f, rsurface.batchtvector3f, rsurface.batchnormal3f, rsurface.batchlightmapcolor4f, rsurface.batchtexcoordtexture2f, rsurface.batchtexcoordlightmap2f);
 		R_SetupShader_SetPermutationSoft(mode, permutation);
@@ -2980,6 +2987,7 @@ void R_SetupShader_Surface(const float rtlightambient[3], const float rtlightdif
 				if (permutation & SHADERPERMUTATION_SHADOWMAPVSDCT    ) R_Mesh_TexBind(GL20TU_CUBEPROJECTION    , r_shadow_shadowmapvsdcttexture                      );
 			}
 		}
+#endif
 		break;
 	}
 }
@@ -3079,6 +3087,7 @@ void R_SetupShader_DeferredLight(const rtlight_t *rtlight)
 	case RENDERPATH_GLES1:
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		R_SetupShader_SetPermutationGLSL(mode, permutation);
 		DPSOFTRAST_Uniform3f(       DPSOFTRAST_UNIFORM_LightPosition            , viewlightorigin[0], viewlightorigin[1], viewlightorigin[2]);
 		DPSOFTRAST_UniformMatrix4fv(DPSOFTRAST_UNIFORM_ViewToLightM1            , 1, false, viewtolight16f);
@@ -3096,6 +3105,7 @@ void R_SetupShader_DeferredLight(const rtlight_t *rtlight)
 		R_Mesh_TexBind(GL20TU_CUBE               , rsurface.rtlight->currentcubemap                    );
 		R_Mesh_TexBind(GL20TU_SHADOWMAP2D        , r_shadow_shadowmap2ddepthtexture                    );
 		R_Mesh_TexBind(GL20TU_CUBEPROJECTION     , r_shadow_shadowmapvsdcttexture                      );
+#endif
 		break;
 	}
 }
@@ -5598,6 +5608,8 @@ void R_SetupView(qboolean allowwaterclippingplane, int fbo, rtexture_t *depthtex
 		R_Viewport_InitPerspective(&r_refdef.view.viewport, &r_refdef.view.matrix, r_refdef.view.x, rtheight - scaledheight - r_refdef.view.y, scaledwidth, scaledheight, r_refdef.view.frustum_x, r_refdef.view.frustum_y, r_refdef.nearclip, r_refdef.farclip, customclipplane);
 	R_Mesh_SetRenderTargets(fbo, depthtexture, colortexture, NULL, NULL, NULL);
 	R_SetViewport(&r_refdef.view.viewport);
+
+#ifdef HAVE_DPSOFTRAST
 	if (r_refdef.view.useclipplane && allowwaterclippingplane && vid.renderpath == RENDERPATH_SOFT)
 	{
 		matrix4x4_t mvpmatrix, invmvpmatrix, invtransmvpmatrix;
@@ -5608,6 +5620,7 @@ void R_SetupView(qboolean allowwaterclippingplane, int fbo, rtexture_t *depthtex
 		Matrix4x4_Transform4(&invtransmvpmatrix, plane, screenplane);
 		DPSOFTRAST_ClipPlane(screenplane[0], screenplane[1], screenplane[2], screenplane[3]);
 	}
+#endif
 }
 
 void R_EntityMatrix(const matrix4x4_t *matrix)
@@ -5643,8 +5656,10 @@ void R_EntityMatrix(const matrix4x4_t *matrix)
 #endif
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_UniformMatrix4fv(DPSOFTRAST_UNIFORM_ModelViewProjectionMatrixM1, 1, false, gl_modelviewprojection16f);
 			DPSOFTRAST_UniformMatrix4fv(DPSOFTRAST_UNIFORM_ModelViewMatrixM1, 1, false, gl_modelview16f);
+#endif
 			break;
 		case RENDERPATH_GL20:
 		case RENDERPATH_GLES2:
@@ -6225,7 +6240,9 @@ static void R_Water_ProcessPlanes(int fbo, rtexture_t *depthtexture, rtexture_t
 		}
 
 	}
+#ifdef HAVE_DPSOFTRAST
 	if(vid.renderpath==RENDERPATH_SOFT) DPSOFTRAST_ClipPlane(0, 0, 0, 1);
+#endif
 	r_fb.water.renderingscene = false;
 	r_refdef.view = originalview;
 	R_ResetViewRendering3D(fbo, depthtexture, colortexture);
@@ -6832,6 +6849,7 @@ static void R_BlendView(int fbo, rtexture_t *depthtexture, rtexture_t *colortext
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			R_Mesh_PrepareVertices_Mesh_Arrays(4, r_screenvertex3f, NULL, NULL, NULL, NULL, r_fb.screentexcoord2f, r_fb.bloomtexcoord2f);
 			R_SetupShader_SetPermutationSoft(SHADERMODE_POSTPROCESS, permutation);
 			R_Mesh_TexBind(GL20TU_FIRST     , r_fb.colortexture);
@@ -6846,6 +6864,7 @@ static void R_BlendView(int fbo, rtexture_t *depthtexture, rtexture_t *colortext
 			DPSOFTRAST_Uniform1f(DPSOFTRAST_UNIFORM_Saturation        , r_glsl_saturation.value);
 			DPSOFTRAST_Uniform2f(DPSOFTRAST_UNIFORM_PixelToScreenTexCoord, 1.0f/vid.width, 1.0f/vid.height);
 			DPSOFTRAST_Uniform4f(DPSOFTRAST_UNIFORM_BloomColorSubtract   , r_bloom_colorsubtract.value, r_bloom_colorsubtract.value, r_bloom_colorsubtract.value, 0.0f);
+#endif
 			break;
 		default:
 			break;
diff --git a/gl_textures.c b/gl_textures.c
index 3f2366b..b5f7667 100644
--- a/gl_textures.c
+++ b/gl_textures.c
@@ -425,8 +425,10 @@ void R_FreeTexture(rtexture_t *rt)
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		if (glt->texnum)
 			DPSOFTRAST_Texture_Free(glt->texnum);
+#endif
 		break;
 	}
 
@@ -624,11 +626,13 @@ static void GL_TextureMode_f (void)
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		// change all the existing texture objects
 		for (pool = gltexturepoolchain;pool;pool = pool->next)
 			for (glt = pool->gltchain;glt;glt = glt->chain)
 				if (glt->texnum && (gl_filter_force || !(glt->flags & (TEXF_FORCENEAREST | TEXF_FORCELINEAR))))
 					DPSOFTRAST_Texture_Filter(glt->texnum, (glt->flags & TEXF_MIPMAP) ? dpsoftrast_filter_mipmap : dpsoftrast_filter_nomipmap);
+#endif
 		break;
 	}
 }
@@ -1241,7 +1245,9 @@ static void R_UploadPartialTexture(gltexture_t *glt, const unsigned char *data,
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		DPSOFTRAST_Texture_UpdatePartial(glt->texnum, 0, data, fragx, fragy, fragwidth, fragheight);
+#endif
 		break;
 	}
 }
@@ -1569,6 +1575,7 @@ static void R_UploadFullTexture(gltexture_t *glt, const unsigned char *data)
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		switch(glt->texturetype)
 		{
 		case GLTEXTURETYPE_2D:
@@ -1619,6 +1626,7 @@ static void R_UploadFullTexture(gltexture_t *glt, const unsigned char *data)
 			DPSOFTRAST_Texture_Filter(glt->texnum, dpsoftrast_filter_mipmap);
 		else
 			DPSOFTRAST_Texture_Filter(glt->texnum, dpsoftrast_filter_nomipmap);
+#endif
 		break;
 	}
 }
@@ -1897,6 +1905,7 @@ static rtexture_t *R_SetupTexture(rtexturepool_t *rtexturepool, const char *iden
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		{
 			int tflags = 0;
 			switch(textype)
@@ -1923,6 +1932,7 @@ static rtexture_t *R_SetupTexture(rtexturepool_t *rtexturepool, const char *iden
 			if (glt->flags & TEXF_CLAMP) tflags |= DPSOFTRAST_TEXTURE_FLAG_CLAMPTOEDGE;
 			glt->texnum = DPSOFTRAST_Texture_New(tflags, glt->tilewidth, glt->tileheight, glt->tiledepth);
 		}
+#endif
 		break;
 	}
 
@@ -2058,6 +2068,7 @@ rtexture_t *R_LoadTextureRenderBuffer(rtexturepool_t *rtexturepool, const char *
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		{
 			int tflags = 0;
 			switch(textype)
@@ -2072,6 +2083,7 @@ rtexture_t *R_LoadTextureRenderBuffer(rtexturepool_t *rtexturepool, const char *
 			}
 			glt->texnum = DPSOFTRAST_Texture_New(tflags, glt->tilewidth, glt->tileheight, glt->tiledepth);
 		}
+#endif
 		break;
 	}
 
@@ -2894,7 +2906,9 @@ rtexture_t *R_LoadTextureDDSFile(rtexturepool_t *rtexturepool, const char *filen
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		glt->texnum = DPSOFTRAST_Texture_New(((glt->flags & TEXF_CLAMP) ? DPSOFTRAST_TEXTURE_FLAG_CLAMPTOEDGE : 0) | (dds_miplevels > 1 ? DPSOFTRAST_TEXTURE_FLAG_MIPMAP : 0), glt->tilewidth, glt->tileheight, glt->tiledepth);
+#endif
 		break;
 	}
 
@@ -2959,12 +2973,14 @@ rtexture_t *R_LoadTextureDDSFile(rtexturepool_t *rtexturepool, const char *filen
 			Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			if (bytesperblock)
 				Con_DPrintf("FIXME SOFT %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 			else
 				DPSOFTRAST_Texture_UpdateFull(glt->texnum, upload_mippixels);
 			// DPSOFTRAST calculates its own mipmaps
 			mip = dds_miplevels;
+#endif
 			break;
 		}
 		if(upload_mippixels != mippixels)
@@ -3040,6 +3056,7 @@ rtexture_t *R_LoadTextureDDSFile(rtexturepool_t *rtexturepool, const char *filen
 		Con_DPrintf("FIXME D3D11 %s:%i %s\n", __FILE__, __LINE__, __FUNCTION__);
 		break;
 	case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 		if (glt->flags & TEXF_FORCELINEAR)
 			DPSOFTRAST_Texture_Filter(glt->texnum, DPSOFTRAST_TEXTURE_FILTER_LINEAR);
 		else if (glt->flags & TEXF_FORCENEAREST)
@@ -3048,6 +3065,7 @@ rtexture_t *R_LoadTextureDDSFile(rtexturepool_t *rtexturepool, const char *filen
 			DPSOFTRAST_Texture_Filter(glt->texnum, dpsoftrast_filter_mipmap);
 		else
 			DPSOFTRAST_Texture_Filter(glt->texnum, dpsoftrast_filter_nomipmap);
+#endif
 		break;
 	}
 
diff --git a/makefile b/makefile
index d18ba6a..77aef24 100644
--- a/makefile
+++ b/makefile
@@ -65,6 +65,18 @@ else
 	CMD_MKDIR=$(CMD_UNIXMKDIR)
 endif
 
+# CPU-specifics
+OBJ_CPU =
+
+ifeq ($(DP_MAKE_TARGET),mingw)
+	# MinGW is always x86
+	OBJ_CPU += $(OBJ_X86)
+else
+ifneq (,$(filter i%86 x86_64,$(DP_MACHINE)))
+	OBJ_CPU += $(OBJ_X86)
+endif
+endif
+
 # 64bits AMD CPUs use another lib directory
 ifeq ($(DP_MACHINE),x86_64)
 	UNIX_X11LIBPATH:=/usr/X11R6/lib64
diff --git a/makefile.inc b/makefile.inc
index de97fb0..74fae2d 100644
--- a/makefile.inc
+++ b/makefile.inc
@@ -98,7 +98,6 @@ OBJ_COMMON= \
 	csprogs.o \
 	curves.o \
 	cvar.o \
-	dpsoftrast.o \
 	dpvsimpledecode.o \
 	filematch.o \
 	fractalnoise.o \
@@ -123,7 +122,6 @@ OBJ_COMMON= \
 	matrixlib.o \
 	mdfour.o \
 	meshqueue.o \
-	mod_skeletal_animatevertices_sse.o \
 	mod_skeletal_animatevertices_generic.o \
 	model_alias.o \
 	model_brush.o \
@@ -163,11 +161,15 @@ OBJ_MENU= \
 	menu.o \
 	mvm_cmds.o
 
+OBJ_X86 = \
+	dpsoftrast.o \
+	mod_skeletal_animatevertices_sse.o
+
 # note that builddate.c is very intentionally not compiled to a .o before
 # being linked, because it should be recompiled every time an executable is
 # built to give the executable a proper date string
-OBJ_SV= builddate.c sys_linux.o vid_null.o thread_null.o $(OBJ_SND_NULL) $(OBJ_COMMON)
-OBJ_SDL= builddate.c sys_sdl.o vid_sdl.o thread_sdl.o $(OBJ_MENU) $(OBJ_SND_COMMON) snd_sdl.o $(OBJ_SDLCD) $(OBJ_VIDEO_CAPTURE) $(OBJ_COMMON)
+OBJ_SV= builddate.c sys_linux.o vid_null.o thread_null.o $(OBJ_SND_NULL) $(OBJ_COMMON) $(OBJ_CPU)
+OBJ_SDL= builddate.c sys_sdl.o vid_sdl.o thread_sdl.o $(OBJ_MENU) $(OBJ_SND_COMMON) snd_sdl.o $(OBJ_SDLCD) $(OBJ_VIDEO_CAPTURE) $(OBJ_COMMON) $(OBJ_CPU)
 
 
 # Compilation
@@ -207,7 +209,7 @@ LDFLAGS_RELEASE=$(OPTIM_RELEASE) -DSVNREVISION=`{ test -d .svn && svnversion; }
 
 ##### UNIX specific variables #####
 
-OBJ_GLX= builddate.c sys_linux.o vid_glx.o thread_pthread.o keysym2ucs.o $(OBJ_MENU) $(OBJ_SOUND) $(OBJ_CD) $(OBJ_VIDEO_CAPTURE) $(OBJ_COMMON)
+OBJ_GLX= builddate.c sys_linux.o vid_glx.o thread_pthread.o keysym2ucs.o $(OBJ_MENU) $(OBJ_SOUND) $(OBJ_CD) $(OBJ_VIDEO_CAPTURE) $(OBJ_COMMON) $(OBJ_CPU)
 
 LDFLAGS_UNIXCOMMON=-lm $(LIB_ODE) $(LIB_Z) $(LIB_JPEG) $(LIB_CRYPTO) $(LIB_CRYPTO_RIJNDAEL) $(LIB_PNG) $(LIB_CURL) $(LIB_FREETYPE2)
 LDFLAGS_UNIXCL=-L$(UNIX_X11LIBPATH) -lX11 -lXpm -lXext -lXxf86vm -pthread $(LIB_SOUND) $(LIB_VORBIS) $(LIB_THEORA)
@@ -247,7 +249,7 @@ LDFLAGS_MACOSXCL=$(LDFLAGS_UNIXCOMMON) -ldl -framework IOKit -framework Carbon $
 LDFLAGS_MACOSXSV=$(LDFLAGS_UNIXCOMMON) -ldl
 LDFLAGS_MACOSXSDL=$(LDFLAGS_UNIXCOMMON) -ldl -framework IOKit $(SDLCONFIG_STATICLIBS) ../../../SDLMain.m
 
-OBJ_AGL= builddate.c sys_linux.o vid_agl.o thread_null.o $(OBJ_MENU) $(OBJ_SOUND) $(OBJ_CD) $(OBJ_VIDEO_CAPTURE) $(OBJ_COMMON)
+OBJ_AGL= builddate.c sys_linux.o vid_agl.o thread_null.o $(OBJ_MENU) $(OBJ_SOUND) $(OBJ_CD) $(OBJ_VIDEO_CAPTURE) $(OBJ_COMMON) $(OBJ_CPU)
 
 EXE_MACOSXCL=darkplaces-agl
 EXE_MACOSXCLNEXUIZ=nexuiz-agl
@@ -278,7 +280,7 @@ LDFLAGS_BSDSDL=$(LDFLAGS_UNIXCOMMON) $(LDFLAGS_UNIXSDL)
 
 WINDRES ?= windres
 
-OBJ_WGL= builddate.c sys_win.o vid_wgl.o thread_null.o $(OBJ_MENU) $(OBJ_SND_WIN) $(OBJ_WINCD) $(OBJ_VIDEO_CAPTURE) $(OBJ_COMMON)
+OBJ_WGL= builddate.c sys_win.o vid_wgl.o thread_null.o $(OBJ_MENU) $(OBJ_SND_WIN) $(OBJ_WINCD) $(OBJ_VIDEO_CAPTURE) $(OBJ_COMMON) $(OBJ_CPU)
 
 # Link
 # see LDFLAGS_WINCOMMON in makefile
diff --git a/quakedef.h b/quakedef.h
index c374573..7f4767b 100644
--- a/quakedef.h
+++ b/quakedef.h
@@ -518,6 +518,11 @@ qboolean Sys_HaveSSE2(void);
 
 #include "palette.h"
 
+#ifdef SSE_POSSIBLE
+/* The software rasterizer requires SSE */
+# define HAVE_DPSOFTRAST
+#endif
+
 /// incremented every frame, never reset
 extern int host_framecount;
 /// not bounded in any way, changed at start of every frame, never reset
diff --git a/vid_glx.c b/vid_glx.c
index 4c5c9c9..57379d8 100644
--- a/vid_glx.c
+++ b/vid_glx.c
@@ -674,6 +674,7 @@ static void HandleEvents(void)
 				else
 					Con_DPrintf("Updating to ConfigureNotify resolution %dx%d\n", vid.width, vid.height);
 
+#ifdef HAVE_DPSOFTRAST
 				if(vid.renderpath == RENDERPATH_SOFT)
 				{
 					DPSOFTRAST_Flush();
@@ -687,6 +688,7 @@ static void HandleEvents(void)
 					vid.softpixels = (unsigned int *) vidx11_ximage[vidx11_ximage_pos]->data;
 					vid.softdepthpixels = (unsigned int *)calloc(4, vid.width * vid.height);
 				}
+#endif
 			}
 			break;
 		case DestroyNotify:
@@ -909,6 +911,7 @@ void VID_Finish (void)
 	switch(vid.renderpath)
 	{
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			if(vidx11_shmevent >= 0) {
 				vidx11_ximage_pos = !vidx11_ximage_pos;
 				vid.softpixels = (unsigned int *) vidx11_ximage[vidx11_ximage_pos]->data;
@@ -931,6 +934,7 @@ void VID_Finish (void)
 				DPSOFTRAST_Finish();
 				XPutImage(vidx11_display, win, vidx11_gc, vidx11_ximage[vidx11_ximage_pos], 0, 0, 0, 0, vid.width, vid.height);
 			}
+#endif
 			break;
 
 		case RENDERPATH_GL11:
@@ -1003,6 +1007,7 @@ static void VID_BuildGLXAttrib(int *attrib, qboolean stencil, qboolean stereobuf
 	*attrib++ = None;
 }
 
+#ifdef HAVE_DPSOFTRAST
 static qboolean VID_InitModeSoft(viddef_mode_t *mode)
 {
 	int i, j;
@@ -1307,6 +1312,7 @@ static qboolean VID_InitModeSoft(viddef_mode_t *mode)
 
 	return true;
 }
+#endif
 
 static qboolean VID_InitModeGL(viddef_mode_t *mode)
 {
@@ -1655,7 +1661,7 @@ static qboolean VID_InitModeGL(viddef_mode_t *mode)
 
 qboolean VID_InitMode(viddef_mode_t *mode)
 {
-#ifdef SSE_POSSIBLE
+#ifdef HAVE_DPSOFTRAST
 	if (vid_soft.integer)
 		return VID_InitModeSoft(mode);
 	else
diff --git a/vid_sdl.c b/vid_sdl.c
index 10a3a50..921e423 100644
--- a/vid_sdl.c
+++ b/vid_sdl.c
@@ -2719,6 +2719,7 @@ extern cvar_t gl_info_version;
 extern cvar_t gl_info_platform;
 extern cvar_t gl_info_driver;
 
+#ifdef HAVE_DPSOFTRAST
 static qboolean VID_InitModeSoft(viddef_mode_t *mode)
 {
 #if SDL_MAJOR_VERSION == 1
@@ -2824,6 +2825,7 @@ static qboolean VID_InitModeSoft(viddef_mode_t *mode)
 #endif
 	return true;
 }
+#endif
 
 qboolean VID_InitMode(viddef_mode_t *mode)
 {
@@ -2837,7 +2839,7 @@ qboolean VID_InitMode(viddef_mode_t *mode)
 #if SDL_MAJOR_VERSION != 1
 	Cvar_SetValueQuick(&vid_touchscreen_supportshowkeyboard, SDL_HasScreenKeyboardSupport() ? 1 : 0);
 #endif
-#ifdef SSE_POSSIBLE
+#ifdef HAVE_DPSOFTRAST
 	if (vid_soft.integer)
 		return VID_InitModeSoft(mode);
 	else
@@ -2930,6 +2932,7 @@ void VID_Finish (void)
 #endif
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_Finish();
 #if SDL_MAJOR_VERSION == 1
 //		if (!r_test.integer)
@@ -2943,6 +2946,7 @@ void VID_Finish (void)
 				SDL_BlitSurface(vid_softsurface, NULL, screen, NULL);
 				SDL_UpdateWindowSurface(window);
 			}
+#endif
 #endif
 			break;
 		case RENDERPATH_D3D9:
diff --git a/vid_shared.c b/vid_shared.c
index 77392a8..276b99d 100644
--- a/vid_shared.c
+++ b/vid_shared.c
@@ -1636,7 +1636,7 @@ static dllhandle_t xinputdll_dll = NULL;
 
 void VID_Shared_Init(void)
 {
-#ifdef SSE_POSSIBLE
+#ifdef HAVE_DPSOFTRAST
 	if (Sys_HaveSSE2())
 	{
 		Con_Printf("DPSOFTRAST available (SSE2 instructions detected)\n");
@@ -2027,6 +2027,7 @@ size_t VID_SortModes(vid_mode_t *modes, size_t count, qboolean usebpp, qboolean
 	return count;
 }
 
+#ifdef HAVE_DPSOFTRAST
 void VID_Soft_SharedSetup(void)
 {
 	gl_platform = "DPSOFTRAST";
@@ -2098,3 +2099,4 @@ void VID_Soft_SharedSetup(void)
 	// clear to black (loading plaque will be seen over this)
 	GL_Clear(GL_COLOR_BUFFER_BIT, NULL, 1.0f, 128);
 }
+#endif
diff --git a/vid_wgl.c b/vid_wgl.c
index e6dee11..d509d35 100644
--- a/vid_wgl.c
+++ b/vid_wgl.c
@@ -307,11 +307,13 @@ void VID_Finish (void)
 		case RENDERPATH_D3D11:
 			break;
 		case RENDERPATH_SOFT:
+#ifdef HAVE_DPSOFTRAST
 			DPSOFTRAST_Finish();
 //			baseDC = GetDC(mainwindow);
 			BitBlt(baseDC, 0, 0, vid.width, vid.height, vid_softhdc, 0, 0, SRCCOPY);
 //			ReleaseDC(mainwindow, baseDC);
 //			baseDC = NULL;
+#endif
 			break;
 		}
 	}
@@ -1562,6 +1564,7 @@ qboolean VID_InitModeDX(viddef_mode_t *mode, int version)
 }
 #endif
 
+#ifdef HAVE_DPSOFTRAST
 qboolean VID_InitModeSOFT(viddef_mode_t *mode)
 {
 	int i;
@@ -1815,10 +1818,11 @@ qboolean VID_InitModeSOFT(viddef_mode_t *mode)
 
 	return true;
 }
+#endif
 
 qboolean VID_InitMode(viddef_mode_t *mode)
 {
-#ifdef SSE_POSSIBLE
+#ifdef HAVE_DPSOFTRAST
 	if (vid_soft.integer)
 		return VID_InitModeSOFT(mode);
 #endif
